[FLOC-958] Add DigitalOcean acceptance test runner

wallrj opened the pull request:
Fixes https://clusterhq.atlassian.net/browse/FLOC-958

wallrj commented:
Hmm. I think I'm going to change tack.
I'm ending up writing a new DigitalOceanV2 API and accompanying tests....which is frustrating, and turning this into a monster branch...and I'm getting bogged down.

I think instead, I might see if I can use an existing Python wrapper to make the v2 API calls when necessary:
 * https://github.com/koalalorenzo/python-digitalocean seems to be the most actively maintained, but it's GPL....which may rule it out.
 * https://github.com/valerylisay/digitalocean-api seems less well maintained and less popular but it's MIT licensed.

wallrj commented on docs/gettinginvolved/acceptance-testing.rst:
```diff
@@ -133,3 +133,27 @@ You will need a ssh agent running with access to the corresponding private key.
+     client_id: "<DigitalOcean API v1 client id>"
+     api_key: "<DigitalOcean API v1 api key>"
+     token: "<DigitalOcean API v2 api token>"
+     location_id: "<DigitalOcean location ID e.g. 9 = ams3>"
```
Figure out whether we can supply a location `slug` instead  eg `lon1`

wallrj commented:
I'm ready for some feedback on this.

I've been running it like so `./admin/run-acceptance-tests --distribution=fedora-20 --provider=digitalocean --config-file=digitalocean.config --keep --flocker-version=0.3.2-1109-gd71621f --branch=digitalocean-acceptance-FLOC-958`

There are lots of rough edges, but I'm interested to hear what @tomprince thinks.

tomprince commented on docs/gettingstarted/installation.rst:
```diff
@@ -263,19 +263,18 @@ You'll probably want to setup at least two nodes.
 
+   #. Upgrade the kernel package inside the virtual machine:
+
+      This specific kernel is no-longer available from the standard Fedora 20 repositories, so we install from the ``kojipkgs`` repository directly.
```
I'd just say "from koij". It isn't a yum repository, and the service is called koji.

tomprince commented:
A brief scan suggests this look reasonable.

adamtheturtle commented on flocker/provision/test/test_digitalocean.py:
```diff
@@ -0,0 +1,64 @@
+        """
+        Set up a test droplet and destroy it after the test.
+        """
+        import pyocean
```
:question: Use absolute imports at the top of the file.

adamtheturtle commented on setup.py:
```diff
@@ -185,6 +185,7 @@ def run(self):
             # The cloud acceptance test runner needs these
             "fabric==1.10.0",
             "apache-libcloud==0.16.0",
+            "digitalocean-python==0.1.5",
```
New dependencies should go in the internal document with their licences. It might be helpful for future contributors if this is mentioned here in `setup.py`.

wallrj replied:
Done.

adamtheturtle commented on flocker/provision/test/test_digitalocean.py:
```diff
@@ -0,0 +1,64 @@
+        # Need to query again for the droplet after updating its kernel
+        updated_droplet = self.client.droplet.get(self.droplet.id)
+
+        # pyocean wraps kernel attributes in its ``Image`` class...which makes
```
Raise an issue upstream and link to it here.

adamtheturtle commented on flocker/provision/test/test_digitalocean.py:
```diff
@@ -0,0 +1,64 @@
+        """
+        import pyocean
+
+        token = os.environ.get('DIGITALOCEAN_TOKEN')
```
Document setting this environment variable.

wallrj replied:
Done.

adamtheturtle commented on docs/gettinginvolved/infrastructure/release-process.rst:
```diff
@@ -409,27 +409,18 @@ Post-Release Review Process
-        vagrant init clusterhq/fedora20-updated
-        vagrant up
-        vagrant ssh
+     The acceptance tests can be run automatically using the ``run-acceptance-tests`` tool.
```
`s/automatically\ //`

adamtheturtle commented on docs/gettingstarted/installation.rst:
```diff
@@ -263,19 +263,18 @@ You'll probably want to setup at least two nodes.
    At the time of writing, the only supported Fedora 20 kernel on DigitalOcean is ``Fedora 20 x64 vmlinuz-3.16.6-203.fc20.x86_64``.
```
It seems that `Fedora 20 x64 vmlinuz-3.17.8-200.fc20.x86_64` is chosen by `set_latest_droplet_kernel` so one of the two is wrong. I guess that this document needs updating.

wallrj replied:
Yeah, well since our automated tests now select what ever is currently the newest kernel, I've updated the documentation along the lines of  "choose the newest kernel".

adamtheturtle commented on docs/gettingstarted/installation.rst:
```diff
@@ -263,19 +263,18 @@ You'll probably want to setup at least two nodes.
    #. Configure the Droplet to boot with the desired kernel:
 
       * Go to the DigitalOcean control panel for your specific Droplet, and in the Settings section choose the Kernel tab.
       * Choose the ``Fedora 20 x64 vmlinuz-3.16.6-203.fc20.x86_64`` kernel for Fedora 20 (scroll all the way to the bottom) and press "Change".
```
Ditto https://github.com/ClusterHQ/flocker/pull/1043/files#r23073958.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,216 @@
+    )
+
+    # libcloud doesn't support shutting down DO vms.
+    # XXX Create an issue.
```
Yes, do that :smile:

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,216 @@
+
+    # libcloud doesn't support powering up DO vms.
+    # XXX Create an issue.
+    retry_if_pending(v2droplet.power_on)
```
https://github.com/ClusterHQ/flocker/pull/1043/files#r23074028

adamtheturtle commented:
This is in review but there are a few XXX's around. I'll stop commenting on them and just say that they should all be addressed before the next submission.

adamtheturtle commented on admin/acceptance.py:
```diff
@@ -146,7 +147,7 @@ def start_nodes(self):
         metadata.update(self.metadata)
 
-        for index in range(2):
+        for index in range(1):
```
I will run this and check but it seems that this is changing from creating two acceptance testing nodes to one.

wallrj replied:
That's just to speed things up while I'm working on the DigitalOcean parts. I'll revert once the branch is ready.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,223 @@
+            return result
+
+
+def set_latest_droplet_kernel(droplet, kernel_prefix='Fedora 20 x64'):
```
It seems that the kernel_prefix parameter is never used so it does not have to be a parameter.

adamtheturtle replied:
Add a note here about the known-working kernel. I imagine that will be useful for debugging if a future kernel causes problems.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,223 @@
+
+def location_by_slug(driver, location_slug):
+    """
+    Look up a DigitalOcean by its short human readable "slug" code.
```
a DigitalOcean ___ ?

adamtheturtle commented on docs/gettinginvolved/acceptance-testing.rst:
```diff
@@ -133,3 +133,27 @@ You will need a ssh agent running with access to the corresponding private key.
+DigitalOcean
+------------
+
+To run the acceptance on DigitalOcean, you need:
```
`sed s/acceptance/acceptance\ tests/`

adamtheturtle replied:
Ah I've just noticed that this seems to be a copy paste error also present in the AWS instructions: "To run the acceptance on AWS, you need". Probably best to fix that too while you're at it.

adamtheturtle commented on docs/gettinginvolved/infrastructure/release-process.rst:
```diff
@@ -409,27 +409,18 @@ Post-Release Review Process
 
      XXX: This step should be automated. See `FLOC-1039 <https://clusterhq.atlassian.net/browse/FLOC-1039>`_.
```
This comment and issue can be modified / removed now, right?

wallrj replied:
That's for automated installation of the client. This is only for node.

adamtheturtle commented on docs/gettinginvolved/infrastructure/release-process.rst:
```diff
@@ -409,27 +409,18 @@ Post-Release Review Process
+     See :ref:acceptance-testing for details of the configuration files required for each platform.
 
-     Follow the :ref:`Flocker node installation documentation<installing-flocker-node>`.
+     .. code-block:: console
```
It isn't clear whether this has to be run from each of the supported client platforms, or just one and the client installation instructions have to be tested on each platform.

adamtheturtle commented on docs/gettinginvolved/infrastructure/release-process.rst:
```diff
@@ -409,27 +409,18 @@ Post-Release Review Process
-     `FLOC-957 <https://clusterhq.atlassian.net/browse/FLOC-957>`_,
-     `FLOC-958 <https://clusterhq.atlassian.net/browse/FLOC-958>`_
-     ).
+        $ ./admin/run-acceptance-tests --distribution=fedora-20 --provider=vagrant --flocker-version=0.3.2-1109-gd71621f --branch=digitalocean-acceptance-FLOC-958
```
Set the `VERSION` environment variable and then set the `--branch` here to be the release branch.

`--flocker-version` can also be `$VERSION`.

adamtheturtle replied:
If you mention that this section, "Post-Release Review Process" requires the setup done in "Pre-tag Review Process" then there is no need to set the VERSION environment variable again and more importantly all of the dependencies needed (e.g. PhantomJS) to run the acceptance tests will already be installed.

wallrj replied:
Well actually I think the version and branch options can be omitted altogether in a real release, because we'll be running the tool from the release branch which will have been tagged, so the version can be discovered automatically.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,224 @@
+
+
+def retry_if_pending(callable, *args, **kwargs):
+    """
```
Document the parameters.

adamtheturtle replied:
:question: You might have an easier time if you use `flocker.testtools.loop_until`.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,224 @@
+    # droplet has halted, but it still seems to require some time before
+    # powering on.
+    # {u'status': u'completed', u'resource_id': 3797602, u'region': u'ams3', u'completed_at': u'2015-01-15T20:52:36Z', u'started_at': u'2015-01-15T20:52:31Z', u'type': u'shutdown', u'id': 41364967, u'resource_type': u'droplet'} # noqa
+    time.sleep(30)
```
Reminder to self: Check if this can be replaced with a `while` loop waiting for something.

wallrj replied:
done.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,224 @@
+    """
+    # Import these here, so that this can be imported without
+    # installing libcloud.
+    from libcloud.compute.providers import get_driver, Provider
```
I may be getting mixed up, but I've seen similar done elsewhere in Flocker with:

```
try:
    from libcloud.compute.providers import get_driver, Provider
    LIBCLOUD_INSTALLED = True
except:
    LIBCLOUD_INSTALLED = False
```

If these have the same consequences, let's try to be consistent. If not, please tell me the differences.

wallrj replied:
I'm just copying the style of other tests. I'll leave it as is.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,224 @@
+    driver = driver_factory(key=client_id, secret=api_key)
+
+    def create_arguments(disk_size):
+        """
```
Explanation of what this does.

adamtheturtle commented on flocker/provision/_install.py:
```diff
@@ -114,6 +114,37 @@ def task_upgrade_kernel():
+
+def koji_kernel_url(version, release, distribution, architecture):
+    """
+    Return the koji URL for the given kernel version.
```
Parameter documentation.

adamtheturtle commented on flocker/provision/_install.py:
```diff
@@ -114,6 +114,37 @@ def task_upgrade_kernel():
+    return url
+
+
+def task_install_kernel(version='3.16.6', release='203', distribution='fc20',
```
This seems to be called only from one place (and that one place specifies these arguments), and so doesn't need these defaults. In particular this is the case because the version specified here is not the one given by the one caller, and so it is never exercised.

adamtheturtle replied:
Ah the defaults seem to be used in `.. task:: install_kernel` in the docs. It might be sensible to note that here. Also, as noted elsewhere, the defaults are now out of date, or at least they don't specify the latest version. It might be too much to find the latest version at run time, but if not, do that. Else update them so they're at least up to date now.

How bad is it to specify an old kernel version here?

adamtheturtle commented on flocker/provision/_install.py:
```diff
@@ -114,6 +114,37 @@ def task_upgrade_kernel():
+
+def task_install_kernel(version='3.16.6', release='203', distribution='fc20',
+                        architecture='x86_64'):
+    """
```
Parameter documentation.

adamtheturtle commented on docs/gettinginvolved/infrastructure/release-process.rst:
```diff
@@ -409,27 +409,18 @@ Post-Release Review Process
-        vagrant up
-        vagrant ssh
+     The acceptance tests can be run automatically using the ``run-acceptance-tests`` tool.
+     See :ref:acceptance-testing for details of the configuration files required for each platform.
```
This renders as:

```
See :ref:acceptance-testing for details of the configuration files required for each platform.
```

So the reference doesn't work.

adamtheturtle replied:
:question: Perhaps instead of linking to the document, which starts with an unnecessary-for-this explanation of how the acceptance test runner works, each needed section could be linked to.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,224 @@
+
+    The raw DO API returns ``event``s whose status can be queried, and that
+    would be a better way to block before issuing the next API call, but
+    pyocean doesn't consistently return the event info. E.g. droplet.create
```
File an issue upstream and link to it here.

wallrj replied:
No, I've spent too long on this now and I'm really sure what to report.

adamtheturtle commented on docs/gettinginvolved/acceptance-testing.rst:
```diff
@@ -133,3 +133,27 @@ You will need a ssh agent running with access to the corresponding private key.
+     client_id: "<DigitalOcean API v1 client id>"
+     api_key: "<DigitalOcean API v1 api key>"
+     token: "<DigitalOcean API v2 api token>"
+     location: "<DigitalOcean location slug e.g. lon1>"
```
How do I best choose the location? I guess I look at a list of slugs somewhere and choose the one which seems closest to me.

wallrj replied:
It's not easy to lookup the slugs except if you use the API or if you manually create a droplet and look at its location in the web interface; that will show up as it's slug rather than full location name. I added some more examples in the document, but I'm not going to document it any further because hopefully this will all be automated soon and the configuration will be stored in LastPass somewhere.

adamtheturtle commented on docs/gettinginvolved/acceptance-testing.rst:
```diff
@@ -133,3 +133,27 @@ You will need a ssh agent running with access to the corresponding private key.
+.. code-block:: yaml
+
+   digitalocean:
+     client_id: "<DigitalOcean API v1 client id>"
```
Having the examples in string quotes is inconsistent with the examples above.

adamtheturtle commented on docs/gettinginvolved/acceptance-testing.rst:
```diff
@@ -133,3 +133,27 @@ You will need a ssh agent running with access to the corresponding private key.
+  (https://cloud.digitalocean.com/generate_api_key),
+- an "API v2" token, which will be used to update the kernel of new droplets,
+  (https://cloud.digitalocean.com/settings/applications), and
+- an ssh-key registered with the AWS account.
```
The DigitalOcean account, right?

adamtheturtle commented on docs/gettinginvolved/acceptance-testing.rst:
```diff
@@ -133,3 +133,27 @@ You will need a ssh agent running with access to the corresponding private key.
+     location: "<DigitalOcean location slug e.g. lon1>"
+     keyname: <ssh-key-name>
+   metadata:
+     creator: <your-name>
```
The other examples have below them a command to run the acceptance tests on that provider using `config.yml`.

adamtheturtle commented on docs/gettinginvolved/infrastructure/release-process.rst:
```diff
@@ -409,27 +409,18 @@ Post-Release Review Process
-     `FLOC-958 <https://clusterhq.atlassian.net/browse/FLOC-958>`_
-     ).
+        $ ./admin/run-acceptance-tests --distribution=fedora-20 --provider=vagrant --flocker-version=0.3.2-1109-gd71621f --branch=digitalocean-acceptance-FLOC-958
+        $ ./admin/run-acceptance-tests --distribution=fedora-20 --provider=aws --config-file=aws.config --flocker-version=0.3.2-1109-gd71621f --branch=digitalocean-acceptance-FLOC-958
```
This instructs a user to create three config files: `aws.config`, `rackspace.config` and `digitalocean.config` which works but is different to the `config.yml` used in the linked to file. That file suggests that it is easiest to use one `config.yml` file with all of the credentials. It may be best to keep in LastPass a complete `config.yml` that the reviewer can download and use (with just a change to "<your-name>" perhaps, and maybe locations if speed is important).

adamtheturtle commented on docs/gettinginvolved/infrastructure/release-process.rst:
```diff
@@ -409,27 +409,18 @@ Post-Release Review Process
-   * Follow the :doc:`../../gettingstarted/tutorial/vagrant-setup` part of the tutorial to make sure that the Vagrant nodes start up correctly.
-   * Follow the :doc:`ELK example documentation<../../gettingstarted/examples/linking>` using a Linux client installation and Rackspace Fedora20 nodes.
 
 #. Merge the release pull request.
```
Similarly to the "Pre-tag Review Process" this should only be done if there are no failures, errors or skips in running the acceptance tests, and a course of action should be specified if there is a failure.

adamtheturtle commented on flocker/provision/test/test_install.py:
```diff
@@ -135,3 +135,23 @@ def test_with_branch_and_version(self):
+    """
+    def test_success(self):
+        """
+        ``koji_kernel_url`` returns the expected URL.
```
... when given some known kernel attributes.

adamtheturtle commented on admin/acceptance.py:
```diff
@@ -222,7 +232,7 @@ class RunOptions(Options):
          'One of fedora-20.'],
```
Separate issue, but this is confusing - it has to be `fedora-20` and so the parameter is not currently useful, or this description is wrong? I'll create an issue for either if necessary.

wallrj replied:
I wouldn't bother. We'll soon be adding Ubuntu and Centos so it'll make sense then.

adamtheturtle commented:
I'm not sure that we have a policy on this so perhaps I'm wrong, but I think that you should add that we now have acceptance tests for running Flocker on DigitalOcean to NEWS.

adamtheturtle commented on docs/gettingstarted/installation.rst:
```diff
@@ -293,7 +291,7 @@ You'll probably want to setup at least two nodes.
       .. code-block:: sh
 
          [root@digitalocean]# uname -r
```
This renders as a commented out `uname -r`

wallrj replied:
nothing to do with this branch.

adamtheturtle replied:
https://clusterhq.atlassian.net/browse/FLOC-1235

tomprince replied:
You could just change to
```rst
.. prompt:: bash [root@digitalocean]#

   uname -r
```

adamtheturtle commented on docs/gettingstarted/installation.rst:
```diff
@@ -260,21 +260,19 @@ You'll probably want to setup at least two nodes.
 
-      .. code-block:: sh
+      * Go to the DigitalOcean control panel for your specific Droplet, and in the Settings section choose the Kernel tab.
+      * Choose the newest kernel for Fedora 20 (scroll all the way to the bottom) and press "Change".
```
This has to be the same choice as the koji line below. Either they should both be hardcoded or both tell you to choose / install the latest version somehow.

wallrj replied:
Note to self: Try defining the kernel URL / version using:
 * http://sphinx.readthedocs.org/en/latest/config.html#confval-rst_epilog

wallrj replied:
Done.

adamtheturtle commented on docs/gettinginvolved/infrastructure/release-process.rst:
```diff
@@ -409,27 +409,17 @@ Post-Release Review Process
 
-   * The node package (``flocker-node``) should be installed on all supported platforms.
-     You can use vagrant to boot a clean Fedora 20 machine as follows:
+   * The node package (``flocker-node``) should be installed and tested on all supported platforms.
```
Ideally this will be done by Buildbot, so please create an issue for this and link to it here.

wallrj replied:
Done.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,242 @@
+    # Sorry about this, but shutdown returns the following, indicating that the
+    # droplet has halted, but it still seems to require some time before
+    # powering on.
+    # {u'status': u'completed', u'resource_id': 3797602, u'region': u'ams3', u'completed_at': u'2015-01-15T20:52:36Z', u'started_at': u'2015-01-15T20:52:31Z', u'type': u'shutdown', u'id': 41364967, u'resource_type': u'droplet'} # noqa
```
This can be `pprint`-ed and then you can get rid of the `#noqa`.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,242 @@
+    """
+    Look up a DigitalOcean location by its short human readable "slug" code.
+
+    # XXX: ``libcloud.DigitalOceanDriver.list_locations`` discards the slug
```
Can't the slug just e ignored and the name used?

wallrj replied:
The names are subject to change, I think. SLUG is a reasonable compromise between stability and human readability.

adamtheturtle replied:
:question: add this as a comment

adamtheturtle commented on flocker/provision/test/test_digitalocean.py:
```diff
@@ -0,0 +1,67 @@
+
+TESTING_DROPLET_ATTRIBUTES = {
+    'region': 'lon1',
+    'size': '8gb',
```
Is this just a random choice or have you found, for example, that a 512mb droplet is not suitable?

wallrj replied:
It's the spec that we use in the installation documentation:
 * http://doc-dev.clusterhq.com/gettingstarted/installation.html#using-digitalocean

adamtheturtle commented on flocker/provision/test/test_digitalocean.py:
```diff
@@ -0,0 +1,67 @@
+        """
+        Set up a test droplet and destroy it after the test.
+        """
+        token = os.environ.get('DIGITALOCEAN_TOKEN')
```
:question: Any issue to run this on Buildbot should probably, for the implementer's sake, mention that this is necessary.

adamtheturtle commented:
`installed-package/fedora-20` reports `No module named pyocean`.

adamtheturtle commented on docs/gettinginvolved/acceptance-testing.rst:
```diff
@@ -133,3 +135,27 @@ You will need a ssh agent running with access to the corresponding private key.
+
+- a DigitalOcean account,
+- a "Legacy API v1" Client ID and API key
+  (https://cloud.digitalocean.com/generate_api_key),
```
This gives me a 404.

wallrj replied:
I linked to the v1 API overview page instead.

adamtheturtle commented on docs/gettinginvolved/infrastructure/release-process.rst:
```diff
@@ -409,27 +409,17 @@ Post-Release Review Process
-        vagrant up
-        vagrant ssh
+     The acceptance tests can be run using the ``run-acceptance-tests`` tool.
+     See :ref:`acceptance-testing` for details of the configuration files required for each platform.
```
Note that for DO at least the particular services+something account which can be used.

adamtheturtle commented on docs/gettinginvolved/acceptance-testing.rst:
```diff
@@ -133,3 +135,27 @@ You will need a ssh agent running with access to the corresponding private key.
+  (https://cloud.digitalocean.com/generate_api_key),
+- an "API v2" token, which will be used to update the kernel of new droplets,
+  (https://cloud.digitalocean.com/settings/applications), and
+- an SSH key registered with the DigitalOcean account.
```
link to https://cloud.digitalocean.com/ssh_keys

adamtheturtle commented on docs/gettinginvolved/acceptance-testing.rst:
```diff
@@ -133,3 +135,27 @@ You will need a ssh agent running with access to the corresponding private key.
+     location: <DigitalOcean location slug e.g. lon1, nyc2, or sfo1>
+     keyname: <ssh-key-name>
+   metadata:
+     creator: <your-name>
```
It seems that I can't have a space in my name here.

adamtheturtle replied:
Same goes for unicode characters.

wallrj replied:
I've added validation to only allow alphanumeric creator name.

adamtheturtle replied:
:question: Note here that it must be alphanumeric

wallrj replied:
Not necessary.

adamtheturtle replied:
I've been using " :question: " to mean "not necessary".

adamtheturtle commented:
Unfortunately after about an hour of running this (`./admin/run-acceptance-tests --distribution=fedora-20 --provider=digitalocean --config-file=/Users/Adam/Desktop/digitalocean.config --branch=digitalocean-acceptance-FLOC-958 --flocker-version=0.3.2-1161-gd337ea7`) I ran into https://clusterhq.atlassian.net/browse/FLOC-1184 :cry: 

I'll take the offending line out and try again after the weekend.

adamtheturtle commented:
I just got `fabric.exceptions.NetworkError: Timed out trying to connect to 178.62.112.70 (tried 24 times)`.

adamtheturtle commented:
Thanks @wallrj . I haven't successfully run these tests, which means I can't give the go ahead. Re-submit after addressing the remaining comments and seeing all the tests pass - getting rid of the intermittent `fabric.exceptions.NetworkError`s. Please also check the comments which GitHub marks as outdated as I think that work done during the review may have caused still relevant comments to be marked as outdated.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,356 @@
+    """
+    Check for a droplet still on exception.
+
+    Sorry about this, but shutdown returns the following, indicating that the
```
I think you were sorry about it being `time.sleep(30)` which it isn't any more.

wallrj commented:
Ready for another review. 

I addressed / answered all the feedback I think.

Can't test until someone fixes our digitalocean testing account.

Not happy with the adding `vmlinuz` and `fc` to the spelling list, but I can't find a way to tell the spell checker to ignore the value of the `digitalocean_kernel_title` substitution. (If I surround it in ````, there's no substitution)

tomprince commented on docs/conf.py:
```diff
@@ -312,3 +314,9 @@ def word_filter(self, word):
+# Define some custom global variables
+# See http://sphinx.readthedocs.org/en/latest/config.html#confval-rst_epilog
+rst_epilog = """
+.. |digitalocean_kernel_title| replace:: {}
```
Would adding `` here address the spelling issue?

tomprince replied:
Around the `{}`.

wallrj replied:
Thanks @tomprince! That works.

adamtheturtle commented:
Fix linting issues (http://build.clusterhq.com/builders/flocker-lint/builds/3672/steps/lint/logs/stdio).

wallrj commented:
This is turning into a nightmare! Apparently DigitalOcean re-assign you the IP addresses that you've previously used which is causing me ssh known_hosts MITM errors.

```
flocker.acceptance.test_postgres
  PostgresTests
    test_deploy ... @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@    WARNING: REMOTE HOST IDENTIFICATION HAS CHANGED!     @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
IT IS POSSIBLE THAT SOMEONE IS DOING SOMETHING NASTY!
Someone could be eavesdropping on you right now (man-in-the-middle attack)!
It is also possible that a host key has just been changed.
The fingerprint for the ECDSA key sent by the remote host is
e1:5e:32:02:7c:52:12:f8:5f:01:a7:b2:db:99:54:d6.
Please contact your system administrator.
Add correct host key in /home/richard/.ssh/known_hosts to get rid of this message.
Offending ECDSA key in /home/richard/.ssh/known_hosts:125
ECDSA host key for 178.62.119.137 has changed and you have requested strict checking.
Host key verification failed.
                                                    [ERROR]
```

...I guess the same can happen with any of our cloud providers though.

Perhaps I'll create a followup issue to check / ignore the known hosts file for these tests.

exarkun commented:
You could set ... `HOME` ... in the `ssh` process's environment ... and put a `.ssh/ssh_config` there that uses a blank known_hosts file?  Or sets options to not to server key checking.

wallrj commented:
After blanking my known_hosts file it now fails differently and sooner. This was working earlier today :cry: 

```
(958)(digitalocean-acceptance-FLOC-958 ? =)[~/.../HybridLogic/flocker]$ ./admin/run-acceptance-tests --distribution=fedora-20 --provider=digitalocean --config-file=digitalocean.config --flocker-version=0.3.2-1621-g67230fe --branch=digitalocean-acceptance-FLOC-958 --keep
Creating node 0: acceptance-test-richardw-0
[root@178.62.77.65] run: yum update -y https://kojipkgs.fedoraproject.org/packages/kernel/3.17.8/200.fc20/x86_64/kernel-3.17.8-200.fc20.x86_64.rpm
--keep specified, not destroying nodes.
Traceback (most recent call last):
  File "./admin/run-acceptance-tests", line 13, in <module>
    main(sys.argv[1:], top_level=TOPLEVEL, base_path=BASEPATH)
  File "/home/richard/projects/HybridLogic/flocker/admin/acceptance.py", line 385, in main
    nodes = runner.start_nodes()
  File "/home/richard/projects/HybridLogic/flocker/admin/acceptance.py", line 213, in start_nodes
    node.provision(package_source=self.package_source)
  File "/home/richard/projects/HybridLogic/flocker/flocker/provision/_libcloud.py", line 178, in provision
    distribution=self.distribution,
  File "/home/richard/projects/HybridLogic/flocker/flocker/provision/_digitalocean.py", line 233, in provision_digitalocean
    commands=task_install_digitalocean_kernel()
  File "/home/richard/projects/HybridLogic/flocker/flocker/provision/_install.py", line 92, in run_with_fabric
    handlers[type(command)](command)
  File "/home/richard/projects/HybridLogic/flocker/flocker/provision/_install.py", line 78, in <lambda>
    Run: lambda e: run(e.command),
  File "/home/richard/.virtualenvs/958/lib/python2.7/site-packages/fabric/network.py", line 647, in host_prompting_wrapper
    return func(*args, **kwargs)
  File "/home/richard/.virtualenvs/958/lib/python2.7/site-packages/fabric/operations.py", line 1048, in run
    shell_escape=shell_escape)
  File "/home/richard/.virtualenvs/958/lib/python2.7/site-packages/fabric/operations.py", line 915, in _run_command
    channel=default_channel(), command=wrapped_command, pty=pty,
  File "/home/richard/.virtualenvs/958/lib/python2.7/site-packages/fabric/state.py", line 397, in default_channel
    chan = _open_session()
  File "/home/richard/.virtualenvs/958/lib/python2.7/site-packages/fabric/state.py", line 389, in _open_session
    return connections[env.host_string].get_transport().open_session()
  File "/home/richard/.virtualenvs/958/lib/python2.7/site-packages/fabric/network.py", line 159, in __getitem__
    self.connect(key)
  File "/home/richard/.virtualenvs/958/lib/python2.7/site-packages/fabric/network.py", line 151, in connect
    user, host, port, cache=self, seek_gateway=seek_gateway)
  File "/home/richard/.virtualenvs/958/lib/python2.7/site-packages/fabric/network.py", line 573, in connect
    raise NetworkError(msg, e)
fabric.exceptions.NetworkError: Timed out trying to connect to 178.62.77.65 (tried 24 times)
```

adamtheturtle commented on docs/conf.py:
```diff
@@ -315,3 +317,9 @@ def word_filter(self, word):
+# Define some custom global variables
+# See http://sphinx.readthedocs.org/en/latest/config.html#confval-rst_epilog
+rst_epilog = """
+.. |digitalocean_kernel_title| replace:: ``{}``
```
:question: Note that if you prefer I don't think you need the underscores and can instead have spaces in identifiers. Examples at http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#substitution-definitions.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,370 @@
+    """
+    Repeats the API call if another API event is currently in progress.
+
+    It returns the result if the call eventually succeeds.
```
And if it does not? (I guess a timeout)

wallrj replied:
Yeah, `trial` will eventually halt the test, but it doesn't seem appropriate to mention that here....as discussed.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,370 @@
+    """
+    for ssh_key in driver.ex_list_ssh_keys():
+        if ssh_key.name == ssh_key_name:
+            break
```
:question: This would be more readable if `return ssh_key` too the place of break and then you don't need an `else`.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,370 @@
+    result = driver.connection.request('/sizes')
+    for size_dict in result.object['sizes']:
+        if size_dict['slug'] == size_slug:
+            break
```
:question: This would be more readable if `return driver._to_size(size_dict)` too the place of break and then you don't need an `else`.

adamtheturtle commented on flocker/provision/_libcloud.py:
```diff
@@ -194,6 +194,7 @@ def name(self):
 ], apply_immutable=True)
 class LibcloudProvisioner(object):
     """
+    # XXX: Should be _driver
```
Do this / create a follow up issue.

adamtheturtle commented:
I don't think that the build failures are related but please check.

adamtheturtle commented:
Perhaps getting fabric to try > 24 times might help the intermittent errors.

adamtheturtle commented:
If you need to change `known_hosts` then document that.

adamtheturtle commented on flocker/provision/test/test_digitalocean.py:
```diff
@@ -0,0 +1,117 @@
+    You must supply a DigitalOcean V2 API token by setting
+    ``DIGITALOCEAN_TOKEN`` in the environment before running these tests.
+    """
+    def setUp(self):
```
:question: No point having `setUp` for one test.

adamtheturtle commented on flocker/provision/test/test_digitalocean.py:
```diff
@@ -0,0 +1,117 @@
+    You must supply a DigitalOcean V2 API token by setting
+    ``DIGITALOCEAN_TOKEN`` in the environment before running these tests.
+    """
+    def setUp(self):
```
:question: No point having `setUp` for one test.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,370 @@
+            return result
+
+
+def pending_event(exception):
```
Document :param exception:

wallrj replied:
Done.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,370 @@
+    return False
+
+
+def droplet_still_on(exception):
```
Document :param exception:

wallrj replied:
Done.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,370 @@
+)
+
+
+def kernel_to_digitalocean_version(kernel):
```
Document `kernel` parameter in the :param kernel: way.

wallrj replied:
Done.

adamtheturtle commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,370 @@
+    """
+    Return the newest kernel available for the supplied droplet, with the given
+    distribution and architecture.
+
```
:question: Document raises. Marked optional because I'm not sure if this is something we have to do, but I have seen it around.

adamtheturtle commented:
I get some timeouts when the tests do run. Perhaps increase the timeouts of those tests.

```
flocker.acceptance.test_deployment
  DeploymentTests
    test_application_volume_quotas ...                                     [OK]
    test_application_volume_quotas_changed ...                             [OK]
    test_deploy ...                                                        [OK]
flocker.acceptance.test_environment_variables
  EnvironmentVariableTests
    test_deploy ...                                                        [OK]
    test_environment_variable_used ...                                     [OK]
    test_moving_data ...                                                   [OK]
    test_moving_mysql ...                                                  [OK]
flocker.acceptance.test_linking
  LinkingTests
    test_deploy ...                                                        [OK]
    test_elasticsearch_empty ... No handlers could be found for logger "elasticsearch"
                                          [OK]
    test_kibana_connects_es ...                                            [OK]
    test_logstash_messages_in_elasticsearch ...                         [ERROR]
                        [ERROR]
    test_moving_data ...                                                [ERROR]
                                               [ERROR]
    test_moving_just_elasticsearch ...                                     [OK]
flocker.acceptance.test_moving_applications
  MovingApplicationTests
    test_moving_application ...                                            [OK]
flocker.acceptance.test_moving_data
  MovingDataTests
    test_moving_data ...                                                   [OK]
flocker.acceptance.test_ports
  PortsTests
    test_deployment_with_ports ...                                         [OK]
    test_mongo_accessible ...                                              [OK]
    test_traffic_routed ...                                             [ERROR]
                                            [ERROR]
flocker.acceptance.test_postgres
  PostgresTests
    test_deploy ...                                                        [OK]
    test_moving_postgres ...                                               [OK]
    test_moving_postgres_data ...                                          [OK]

===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.internet.defer.TimeoutError: <flocker.acceptance.test_linking.LinkingTests testMethod=test_logstash_messages_in_elasticsearch> (test_logstash_messages_in_elasticsearch) still running at 120.0 secs

flocker.acceptance.test_linking.LinkingTests.test_logstash_messages_in_elasticsearch
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 0x10ef5dcf8 [0.0123817920685s] called=0 cancelled=0 Deferred.callback(None)>

flocker.acceptance.test_linking.LinkingTests.test_logstash_messages_in_elasticsearch
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.internet.defer.TimeoutError: <flocker.acceptance.test_linking.LinkingTests testMethod=test_moving_data> (test_moving_data) still running at 120.0 secs

flocker.acceptance.test_linking.LinkingTests.test_moving_data
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 0x10f26d128 [0.0633208751678s] called=0 cancelled=0 Deferred.callback(None)>

flocker.acceptance.test_linking.LinkingTests.test_moving_data
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.internet.defer.TimeoutError: <flocker.acceptance.test_ports.PortsTests testMethod=test_traffic_routed> (test_traffic_routed) still running at 120.0 secs

flocker.acceptance.test_ports.PortsTests.test_traffic_routed
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 0x10f2bc6c8 [0.0913870334625s] called=0 cancelled=0 Deferred.callback(None)>

flocker.acceptance.test_ports.PortsTests.test_traffic_routed
-------------------------------------------------------------------------------
Ran 21 tests in 1233.602s

FAILED (errors=6, successes=18)
Destroying acceptance-test-AdamDangoor-0
Destroying acceptance-test-AdamDangoor-1
```

wallrj commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,370 @@
+    :param kwargs: Keyword arguments to supply when calling it.
+    :return: The result of calling  ``callable``.
+    """
+    while True:
```
Can't easily use loop_until because it returns a deferred and all this code is synchronous. But I agree that this is all over complicated. I've simplified things I hope.

wallrj commented on flocker/provision/test/test_digitalocean.py:
```diff
@@ -0,0 +1,117 @@
+}
+
+
+@skipUnless(PYOCEAN_INSTALLED, "pyocean not installed")
```
It raises `SkipTest` which bubbles up and is caught by the test runner. This seemed like a good place for it as it's the only place that actually requires pyocean. On the other hand it would be a bad idea if the function was called after a lengthy setup routine. It's not in this case though so I think I'll leave it.

wallrj commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,370 @@
+
+    It returns the result if the call eventually succeeds.
+
+    :param error_checkers: A ``list`` of ``callables`` which will check for
```
I removed that function.

wallrj commented:
Ok. I addressed most of the feedback from the last review. Ignored some of it.
The tests do now run, but some fail due to timeouts.

```
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.internet.defer.TimeoutError: <flocker.acceptance.test_linking.LinkingTests testMethod=test_logstash_messages_in_elasticsearch> (test_logstash_messages_in_elasticsearch) still running at 120.0 secs

flocker.acceptance.test_linking.LinkingTests.test_logstash_messages_in_elasticsearch
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 0x263a7e8 [0.0789959430695s] called=0 cancelled=0 Deferred.callback(None)>

flocker.acceptance.test_linking.LinkingTests.test_logstash_messages_in_elasticsearch
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.internet.defer.TimeoutError: <flocker.acceptance.test_linking.LinkingTests testMethod=test_moving_data> (test_moving_data) still running at 120.0 secs

flocker.acceptance.test_linking.LinkingTests.test_moving_data
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 0x2987878 [0.0994489192963s] called=0 cancelled=0 Deferred.callback(None)>

flocker.acceptance.test_linking.LinkingTests.test_moving_data
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.internet.defer.TimeoutError: <flocker.acceptance.test_ports.PortsTests testMethod=test_traffic_routed> (test_traffic_routed) still running at 120.0 secs

flocker.acceptance.test_ports.PortsTests.test_traffic_routed
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 0x2a37f38 [0.0993759632111s] called=0 cancelled=0 Deferred.callback(None)>

flocker.acceptance.test_ports.PortsTests.test_traffic_routed
-------------------------------------------------------------------------------
Ran 21 tests in 1094.375s

FAILED (errors=6, successes=18)
--keep specified, not destroying nodes.
```

But we've seen that before and there's now a followup issue to increase the timeouts on acceptance tests:
 * https://clusterhq.atlassian.net/browse/FLOC-1384

I also created a followup issue to deal with SSH known_hosts validation errors:
 * https://clusterhq.atlassian.net/browse/FLOC-1385

adamtheturtle commented:
Did something get changed to avoid `fabric.exceptions.NetworkError: Timed out trying to connect to 178.62.112.70 (tried 24 times)` or is that just luck?

tomprince commented on admin/acceptance.py:
```diff
@@ -220,38 +227,47 @@ def stop_nodes(self):
+def runner_for_cloudprovisioner(provisioner_name):
     """
-    Run the tests against rackspace nodes.
+    Dynamically generate and return a factory function which will be used to
```
This functions feels like a runner factory factory. I think one of the layers should be removed.

wallrj replied:
Done.

tomprince commented on admin/acceptance.py:
```diff
@@ -220,38 +227,47 @@ def stop_nodes(self):
-        raise UsageError("Must provided 'aws' config stanza.")
+        provisioner_factory = namedAny(
+            'flocker.provision.{}_provisioner'.format(
+                provisioner_name))
```
Would it make sense for `flocker.provision` to provide something where we could look up the provisioners by name, rather than using `namedAny`. (Probably just a dictionary somewhere).

wallrj replied:
I added a dictionary of provisioner functions in `flocker.provisioning`.

tomprince commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,373 @@
+from ._common import Kernel
+
+
+def retry_on_error(error_checkers, callable, *args, **kwargs):
```
This should fail eventually.

tomprince replied:
That is, it should fail after some number of iterations, or some amount of time.

tomprince replied:
:question: But, maybe doing this can wait until we convert this to asynchronous code.

tomprince commented on flocker/provision/_common.py:
```diff
@@ -22,3 +22,18 @@ class PackageSource(object):
+)
+class Kernel(object):
+    """
+    Represents the version information for a kernel package.
```
This seems at least slightly RPM specific. Perhaps that should be mentioned here.

wallrj replied:
done.

tomprince commented on flocker/provision/_install.py:
```diff
@@ -114,6 +114,51 @@ def task_upgrade_kernel():
+).format(kernel=DIGITALOCEAN_KERNEL)
+
+
+def task_install_digitalocean_kernel():
```
:question: It would probably make more sense for this to just take a `Kernel` to install. I guess that makes the sphinx more complicated, though.

I was going to suggest that the above to `DIGITIALOCEAN_` constants be moved to `._digitalocean`.

tomprince commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,373 @@
+        raise ValueError("Unknown size slug.", size_slug)
+
+
+def ssh_key_by_name(driver, ssh_key_name):
```
This perhaps belongs beside the `get_image` and `get_size` functions in `._libcloud`.

tomprince commented on flocker/provision/_digitalocean.py:
```diff
@@ -0,0 +1,373 @@
+    :param bytes ssh_key_name: The name of a registered SSH key.
+    :returns: ``SSHKey``
+    """
+    for ssh_key in driver.ex_list_ssh_keys():
```
It makes me very sad that libcloud's digitalocean driver doesn't support the base `list_key_pairs`.

https://issues.apache.org/jira/browse/LIBCLOUD-669

wallrj replied:
So I gather from this, that it's only the DigitalOcean driver that'll need this helper function. So perhaps I'll leave it where it is.

tomprince commented:
There are a couple of suggested code changes, and a merge conflict, so I am going to reject this for the moment. There are also some test failures, but I think they may also trigger on rackspace, so aren't new to this branch.

adamtheturtle commented on docs/gettinginvolved/infrastructure/release-process.rst:
```diff
@@ -510,24 +510,26 @@ Post-Release Review Process
-     You can use vagrant to boot a clean Fedora 20 machine as follows:
+   * The node package (``flocker-node``) should be installed and tested on all supported platforms.
+
+     The acceptance tests can be run using the ``run-acceptance-tests`` tool.
```
Note that there is a race condition here. If https://clusterhq.atlassian.net/browse/FLOC-1304 gets completed before this is merged, this will add unnecessary steps.

tomprince commented on admin/acceptance.py:
```diff
@@ -321,11 +311,10 @@ def postOptions(self):
-        provider_factory = PROVIDERS[self['provider']]
-        self.runner = provider_factory(
-            top_level=self.top_level,
+        self.runner = LibcloudRunner(
```
This appears to drop support for vagrant.

wallrj replied:
Sorry, I missed that. 

I made some more changes which hopefully fix that,.

tomprince commented:
There is a failing acceptance test, but that can be addressed in a follow-up issue.
